bitcoin/src/amount.rs

```diff
@@ -1896,39 +1894,51 @@ mod tests {
-        assert_eq!(sp("-0.1 satoshi"), Err(E::TooPrecise));
-        assert_eq!(p("0.123456 mBTC"), Err(E::TooPrecise));
-        assert_eq!(sp("-1.001 bits"), Err(E::TooPrecise));
-        assert_eq!(sp("-200000000000 BTC"), Err(E::TooBig));
-        assert_eq!(p("18446744073709551616 sat"), Err(E::TooBig));
-
-        assert_eq!(sp("0 msat"), Err(E::TooPrecise));
-        assert_eq!(sp("-0 msat"), Err(E::TooPrecise));
-        assert_eq!(sp("000 msat"), Err(E::TooPrecise));
-        assert_eq!(sp("-000 msat"), Err(E::TooPrecise));
-        assert_eq!(p("0 msat"), Err(E::TooPrecise));
-        assert_eq!(p("-0 msat"), Err(E::TooPrecise));
-        assert_eq!(p("000 msat"), Err(E::TooPrecise));
-        assert_eq!(p("-000 msat"), Err(E::TooPrecise));
-
-        assert_eq!(p(".5 bits"), Ok(Amount::from_sat(50)));
-        assert_eq!(sp("-.5 bits"), Ok(SignedAmount::from_sat(-50)));
-        assert_eq!(p("0.00253583 BTC"), Ok(Amount::from_sat(253583)));
-        assert_eq!(sp("-5 satoshi"), Ok(SignedAmount::from_sat(-5)));
-        assert_eq!(p("0.10000000 BTC"), Ok(Amount::from_sat(100_000_00)));
-        assert_eq!(sp("-100 bits"), Ok(SignedAmount::from_sat(-10_000)));
+
+        assert_eq!(Amount::from_str("x BTC"), Err(E::InvalidCharacter('x')));
+        assert_eq!(Amount::from_str("xBTC"), Err(E::UnknownDenomination("xBTC".into())));
+        assert_eq!(Amount::from_str("5 BTC BTC"), Err(E::UnknownDenomination("BTC BTC".into())));
+        assert_eq!(Amount::from_str("5BTC BTC"), Err(E::InvalidCharacter('B')));
+        assert_eq!(Amount::from_str("5 5 BTC"), Err(E::UnknownDenomination("5 BTC".into())));
+
+        #[track_caller]
+        fn case(s: &str, expected: Result<Amount, ParseAmountError>) {
+          assert_eq!(Amount::from_str(s), expected);
+          assert_eq!(Amount::from_str(&s.replace(" ", "")), expected);
+        }
+
+        #[track_caller]
+        fn scase(s: &str, expected: Result<SignedAmount, ParseAmountError>) {
+          assert_eq!(SignedAmount::from_str(s), expected);
+          assert_eq!(SignedAmount::from_str(&s.replace(" ", "")), expected);
+        }
+
+        case("5 BCH", Err(E::UnknownDenomination("BCH".to_owned())));
+
+        case("-1 BTC", Err(E::Negative));
+        case("-0.0 BTC", Err(E::Negative));
+        case("0.123456789 BTC", Err(E::TooPrecise));
+        scase("-0.1 satoshi", Err(E::TooPrecise));
+        case("0.123456 mBTC", Err(E::TooPrecise));
+        scase("-1.001 bits", Err(E::TooPrecise));
+        scase("-200000000000 BTC", Err(E::TooBig));
+        case("18446744073709551616 sat", Err(E::TooBig));
+
+        scase("0 msat", Err(E::TooPrecise));
+        scase("-0 msat", Err(E::TooPrecise));
+        scase("000 msat", Err(E::TooPrecise));
+        scase("-000 msat", Err(E::TooPrecise));
+        case("0 msat", Err(E::TooPrecise));
+        case("-0 msat", Err(E::TooPrecise));
+        case("000 msat", Err(E::TooPrecise));
+        case("-000 msat", Err(E::TooPrecise));
```

Comment:
These don't look right, 0 msat is just 0. But I guess the bug was there before?

Reply:
Yah, these looked weird to me. I feel like 0 should parse with any denomination, but they were there before so I left them.